### ISSUE TITLE
Include id_token in ResponseType

### ISSUE THREAD

**Issue by BrentSouza** [okta/okta-aspnet#51]
In the AspNet version of the extensions, the ResponseType is set to do code and id_token

```csharp
app.UseOpenIdConnectAuthentication(new OpenIdConnectAuthenticationOptions
{
    ClientId = options.ClientId,
    ClientSecret = options.ClientSecret,
    Authority = issuer,
    RedirectUri = options.RedirectUri,
    ResponseType = OpenIdConnectResponseType.CodeIdToken,
    Scope = scopeString,
    PostLogoutRedirectUri = options.PostLogoutRedirectUri,
    TokenValidationParameters = tokenValidationParameters,
    Notifications = new OpenIdConnectAuthenticationNotifications
    {
        AuthorizationCodeReceived = tokenExchanger.ExchangeCodeForTokenAsync,
        RedirectToIdentityProvider = BeforeRedirectToIdentityProviderAsync,
    },
});
```

However, in the AspNetCore version, it's just code:

```csharp
builder.AddOpenIdConnect(oidcOptions =>
{
    oidcOptions.ClientId = options.ClientId;
    oidcOptions.ClientSecret = options.ClientSecret;
    oidcOptions.Authority = issuer;
    oidcOptions.CallbackPath = new PathString(options.CallbackPath);
    oidcOptions.SignedOutCallbackPath = new PathString(OktaDefaults.SignOutCallbackPath);
    oidcOptions.ResponseType = OpenIdConnectResponseType.Code;
    oidcOptions.GetClaimsFromUserInfoEndpoint = options.GetClaimsFromUserInfoEndpoint;
    oidcOptions.SaveTokens = true;
    oidcOptions.UseTokenLifetime = false;
    oidcOptions.BackchannelHttpHandler = new UserAgentHandler("okta-aspnetcore", typeof(OktaAuthenticationOptionsExtensions).Assembly.GetName().Version);

    var hasDefinedScopes = options.Scope?.Any() ?? false;
    if (hasDefinedScopes)
    {
        oidcOptions.Scope.Clear();
        foreach (var scope in options.Scope)
        {
            oidcOptions.Scope.Add(scope);
        }
    }

    oidcOptions.TokenValidationParameters = new DefaultTokenValidationParameters(options, issuer)
    {
        ValidAudience = options.ClientId,
        NameClaimType = "name",
    };

    oidcOptions.Events.OnRedirectToIdentityProvider = BeforeRedirectToIdentityProviderAsync;
});
```

This prevents returning group claims since the groups scope only works with ID tokens. Is this an intentional change in behavior?

**Comment by laura-rodriguez** [comment]
Hi @BrentSouza,

If you want to include `groups` as part of your claims, then you should go to your `Okta Dashboard > API > Authorization Server`. Once you select the AS you are using go to `Claims` tab and select `Add Claim`.
Select `Groups` in `Value Type` and then configure your desired configuration. 

![image](https://user-images.githubusercontent.com/16751130/47179271-a9a35000-d2eb-11e8-83d6-d8c9d38ff426.png)

In this example, I configured to include this claim as part of `profile` so make sure you include `profile` as part of your `scope`:

```
app.UseOktaMvc(new OktaMvcOptions()
            {
             ...
             Scope = new List<string> {"openid", "profile", "email"},
            });
```
Of course, it depends on how you configure your claim.
The `ResponseType` is related to the `OAuth` flow. If you are using the authorization code flow, you are expected to use just `code`. In `ASP.NET Core` you have access to the tokens thanks to the `SaveTokens` property which is set to `true` when we register the OIDC middleware.
ASP.NET 4.x, is a different story and we have to include `IdToken` in the `ResponseType`.

Let me know if this works for you.

Laura.

**Comment by BrentSouza** [comment]
Thanks that does make sense but we are using the Okta organization auth server and not

**Comment by laura-rodriguez** [comment]
@BrentSouza , what do you mean with a custom auth server? 
Anyways, you could also consider the [Okta .NET SDK](https://github.com/okta/okta-sdk-dotnet) which allows you to manage Groups via the [Groups API](https://developer.okta.com/docs/api/resources/groups).

**Comment by justinhelgerson** [comment]
I *think* the point @BrentSouza is making is that the [middleware for ASP.NET](https://github.com/okta/okta-aspnet/blob/ee8fc237b026b38b4d7e6e0707c3a3447069e2f2/Okta.AspNet/OktaMiddlewareExtensions.cs#L112) is specifying `CodeIdToken` whereas the [middleware for ASP.NET Core](https://github.com/okta/okta-aspnet/blob/ee8fc237b026b38b4d7e6e0707c3a3447069e2f2/Okta.AspNetCore/OktaAuthenticationOptionsExtensions.cs#L46) is only specifying `Code`.

@BrentSouza you can always skip the Okta middleware provided through this library and configure however you need.  Here's an example of what I'm doing:

```
.AddOpenIdConnect(options => {
    options.ClientId = oktaConfig.ClientId;
    options.ClientSecret = oktaConfig.ClientSecret;
    options.Authority = oktaConfig.Authority;
    options.CallbackPath = "/signin-oidc";
    options.ResponseType = OpenIdConnectResponseType.CodeIdToken;
    options.SaveTokens = true;
    options.UseTokenLifetime = false;
    options.GetClaimsFromUserInfoEndpoint = true;
    options.Scope.Add("groups");
    options.Scope.Add("email");
});
```

Then you can configure your app to include the group claims similarly to how @laura-rodriguez described.

**Comment by BrentSouza** [comment]
@laura-rodriguez this is what I am referring to:

https://okta.github.io/docs/how-to/creating-token-with-groups-claim

In the link above the auth servers are referred to by the terminology: Okta and Custom

Right now I don’t have custom authorization servers enabled in my organizations oktapreview account so I am testing with the Okta auth server and manipulating the claims is not exposed.

@justinhelgerson yes that is what I am referring to but I was really asking if the change was intentional. From Laura’s response it seems like it was and I will explore retrieving the id_token from the saved tokens. 

@laura-rodriguez is this what you were referring to by accessing the saved tokens?

```csharp
    string accessToken = await HttpContext.GetTokenAsync("access_token");
    string idToken = await HttpContext.GetTokenAsync("id_token");
```

**Comment by nbarbettini** [comment]
The difference in response_type between the two libraries is just an implementation detail (actually a limitation of the Katana middleware the AspNet library is based on). Both libraries end up with both an access token and an ID token at the end of the login flow. Any scopes in the ID token should be available as claims on the .NET `ClaimsPrincipal`.

Just to double check on your use case @BrentSouza:
1. You're authenticating against the Okta authorization server
2. The ID token returned includes a `groups` claim
3. You want to have access to the contents of the `groups` claim in your AspNetCore code

Does the ID token already have the claim you're expecting in step 2? AspNetCore receives access and ID tokens after it does a code exchange (backchannel), so the tokens won't be visible in the browser. You can use a tool like Fiddler to watch the tokens come back over the wire.

**Comment by BrentSouza** [comment]
@nbarbettini 

> 1. You're authenticating against the Okta authorization server
> 2. The ID token returned includes a `groups` claim
> 3. You want to have access to the contents of the `groups` claim in your AspNetCore code

Yes to all 3.

What happens is that the `ClaimsPrincipal` doesn't contain the `groups` claim. If I get the id_token explicitly and decode the jwt, that also does not have a `groups` claim. I am using the `okta-aspnetcore-mvc-example` project to test things out. Relevant bits below:

`Startup.cs`
```csharp
public void ConfigureServices(IServiceCollection services)
{
    var oktaMvcOptions = new OktaMvcOptions();
    Configuration.GetSection("Okta").Bind(oktaMvcOptions);
    oktaMvcOptions.Scope = new List<string> { "openid", "profile", "email", "groups" };
    oktaMvcOptions.AuthorizationServerId = "";

    services.AddAuthentication(options =>
    {
        options.DefaultAuthenticateScheme = CookieAuthenticationDefaults.AuthenticationScheme;
        options.DefaultSignInScheme = CookieAuthenticationDefaults.AuthenticationScheme;
        options.DefaultChallengeScheme = OktaDefaults.MvcAuthenticationScheme;
    })
    .AddCookie()
    .AddOktaMvc(oktaMvcOptions);

    services.AddMvc();
}
```

`HomeController.cs`
```csharp
[Authorize]
public async Task<IActionResult> Profile()
{
    string idToken = await HttpContext.GetTokenAsync("id_token");
    string accessToken = await HttpContext.GetTokenAsync("access_token");
    return View(HttpContext.User.Claims);
}
```

`HttpContext.User.Claims` has the following claims:

* sub
* name
* email
* jti
* preferred_username
* given_name
* family_name

The decoded id_token has:

* sub
* name
* email
* ver
* iss
* aud
* iat
* exp
* jti
* amr
* idp
* nonce
* preferred_username
* auth_time
* at_hash

I feel like there's just something I'm missing. If I switch to using a hand-rolled `AddOpenIdConnect` similar to what @justinhelgerson wrote above, I do get the `groups` claim in the `ClaimsPrincipal`.

**Comment by nbarbettini** [comment]
@BrentSouza Sorry that we haven't followed up yet! There's something here, but I need to do some more investigation. You're not missing anything obvious 🙂 

In the mean time, are you able to use `AddOpenIdConnect` so you're not blocked?

**Comment by laura-rodriguez** [comment]
Hi @BrentSouza ,
I know it's been a while since you opened this issue (Sorry about that!), but we finally fixed this recently as part of another task.
Now, In ASP.NET Core, if you set the property `OktaMvcOptions.GetClaimsFromUserInfoEndpoint = true` all the claims will be attached to the principal, `groups` among them. 
You can see the details of the implementation [here](https://github.com/okta/okta-aspnet/commit/511072b52e74546edcfe1e59d081b4b399a00903#diff-74c7a610d17f835b558d281cef50447cR70).
Basically, the issue was that in ASP.NET Core only certain claims are attached to the principal by default, that's why many of them weren't present before. 
Thanks for all the info you provided!

**Comment by laura-rodriguez** [comment]
Fixed by #83.

**Comment by VijiMari** [comment]
This might be fixed for ASP.NET Core but the same issue seems to exist in .net framework. When I use okta-aspnet-mvc-example and add groups in the scope as shown below,

namespace okta_aspnet_mvc_example
{
    public class Startup
    {
        public void Configuration(IAppBuilder app)
        {
            app.SetDefaultSignInAsAuthenticationType(CookieAuthenticationDefaults.AuthenticationType);

            app.UseCookieAuthentication(new CookieAuthenticationOptions());

            app.UseOktaMvc(new OktaMvcOptions()
            {
                OktaDomain = ConfigurationManager.AppSettings["okta:OktaDomain"],
                ClientId = ConfigurationManager.AppSettings["okta:ClientId"],
                ClientSecret = ConfigurationManager.AppSettings["okta:ClientSecret"],
                RedirectUri = ConfigurationManager.AppSettings["okta:RedirectUri"],
                PostLogoutRedirectUri = ConfigurationManager.AppSettings["okta:PostLogoutRedirectUri"],
                GetClaimsFromUserInfoEndpoint = true,
                Scope = new List<string> {"openid", "groups", "profile", "email"},
            });
        }
    }
}

it actually throws below error,

"OpenIdConnectMessage.Error was not null, indicating an error. Error: 'invalid_scope'. Error_Description (may be empty): 'One or more scopes are not configured for the authorization server resource.'. Error_Uri (may be empty): ''."

**Comment by laura-rodriguez** [comment]
Hi @VijiMari ,
What authorization server are you using? 
An Okta Org Authorization Server authorization which endpoint looks like this:
`https://${yourOktaDomain}/oauth2/v1/authorize`

Or, a custom authorization server which endpoint looks like this:
`https://${yourOktaDomain}/oauth2/${authServerId}/v1/authorize`

You can find more info about authorization servers here: https://developer.okta.com/docs/guides/customize-authz-server/overview/

**Comment by VijiMari** [comment]
@laura-rodriguez -  'm using Okta Org Authorization Server authorization

**Comment by laura-rodriguez** [comment]
have you created a groups claim in your application? In case you haven't, you can check how to do it here: https://developer.okta.com/docs/guides/customize-tokens-returned-from-okta/-/add-groups-claim-org-as/

**Comment by VijiMari** [comment]
thanks @laura-rodriguez. Found the missing bit.. its actually groups scope in Authorization servers,

![image](https://user-images.githubusercontent.com/26272241/71681108-bf335700-2d83-11ea-9707-36d1a31a190a.png)
